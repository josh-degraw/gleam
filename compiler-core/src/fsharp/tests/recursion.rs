use crate::assert_fsharp;

#[test]
fn tco() {
    assert_fsharp!(
        r#"
pub fn main(x) {
  case x {
    0 -> Nil
    _ -> main(x - 1)
  }
}
"#
    );
}

#[test]
fn tco_case_block() {
    assert_fsharp!(
        r#"
pub fn main(x) {
  case x {
    0 -> Nil
    _ -> {
      let y = x
      main(y - 1)
    }
  }
}
"#
    );
}

// https://github.com/gleam-lang/gleam/issues/1779
#[test]
fn not_tco_due_to_assignment() {
    assert_fsharp!(
        r#"
pub fn main(x) {
  let z = {
    let y = x
    main(y - 1)
  }
  z
}
"#
    );
}

// https://github.com/gleam-lang/gleam/issues/2400
#[test]
fn shadowing_so_not_recursive() {
    // This funtion is calling an argument with the same name as itself, so it is not recursive
    assert_fsharp!(
        r#"
pub fn map(map) {
  map()
}
"#
    );
}
