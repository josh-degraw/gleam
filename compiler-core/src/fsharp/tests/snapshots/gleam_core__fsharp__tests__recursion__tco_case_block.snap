---
source: compiler-core/src/fsharp/tests/recursion.rs
assertion_line: 19
expression: "\npub fn main(x) {\n  case x {\n    0 -> Nil\n    _ -> {\n      let y = x\n      main(y - 1)\n    }\n  }\n}\n"
---
module rec my.``mod``
let main (x: int): unit = begin
    match x with
    | 0 ->
        ()
    | _ ->
        let y = x
        main (y - 1)
end
